### ISSUE TITLE
suspenders should generate coffeescript

### ISSUE THREAD

**Issue by lexi-lambda** [philosophie/suspenders#72]
We should generate an `application.coffee` file instead of `application.js`.

**Comment by taboularasa** [comment]
seems this change should be deferred until we decide on the coffeescript / es6 issue

**Comment by jdwolk** [comment]
@taboularasa I sort of disagree -- I think right now our default is coffeescript, and it's annoying to have to change the default `application.js` -> `application.coffee` on every project. Think it makes sense to bring in and revisit if / when we decide to switch to es6

**Comment by taboularasa** [comment]
@philosophie/devs this issue is implicitly about switching from coffee to es6 as default for new rails projects, so we should put the issue to a vote here and now. ok go:

**Comment by wykhuh** [comment]
ES6. 
- Many of the javascript people I follow on twitter advocate the use of ES6. Saw a few tweets about people using Babel to experiment with ES7 features.
-  A lot of big organizations are using ES6 http://babeljs.io/users/
- We keep up to date with the latest versions of Ruby and Rails. It seems odd to not stay up to date with the latest version of Javascript, especially if we are going to do more javascript projects. 
- Babel community is more active than Coffeescript community. Github stats for the last 30 day.
  Coffeescript: 5 authors have pushed 21 commits to master and 25 commits to all branches. 
  Babel: 18 authors have pushed 19 commits to master and 137 commits to all branches
- Thoughtbot published a post "Replace CoffeeScript with ES6".  https://robots.thoughtbot.com/replace-coffeescript-with-es6 
  From what I can tell, they haven't completely dropped coffeescript, but are advocating both coffeescript and ES6. JS used to be turned off by default in Suspenders. Now JS and Coffeescript are turned on so people can decide which to use. Their style guide advocates Coffeescript or ES6.

**Comment by cltatman** [comment]
sounds to me like es6 vs coffeescript is mostly a wash, lacking motivation to switch from coffeescript. if our go-to in JS land is going to be angular, does anyone have strong feelings about angular + coffeescript vs angular + es6?

**Comment by lexi-lambda** [comment]
> sounds to me like es6 vs coffeescript is mostly a wash

I agree with this. I know I've made my preference for CoffeeScript clear, but I don't see any objective reasons for one over the other. I consider the "it's the future!" argument a weak one, but it's at least somewhat valid. I'm not going to argue heavily for one or the other.

> if our go-to in JS land is going to be angular, does anyone have strong feelings about angular + coffeescript vs angular + es6?

I personally find working with Angular _marginally_ easier in CoffeeScript, but the difference is minimal. The biggest advantage I've found with CoffeeScript is its `@` shorthand for `this`—idiomatic Angular makes frequent use of `this` due to the "controller as" syntax. This is, of course, still pretty trivial. Otherwise, ES6 makes up for ES5's shortcomings when working with Angular. Fat arrow (`=>`) functions are _extremely_ useful when working with Angular due to the aforementioned reason, but ES6 has these now, so nobody wins.

If we really wanted to pick a language best for working with Angular, TypeScript would need to be on the table. All of Angular 2 is built in TypeScript, and the Angular team has a close integration with the TypeScript toolchain. In fact, the syntax for Angular 2 without TypeScript isn't even decided yet, and it currently looks incredibly ugly. Of course, this point doesn't matter for Angular 1.x, which was designed with JavaScript in mind.

tl;dr: I don't care, and I think the objective differences are minimal, but I have a subjective preference for CoffeeScript.

**Comment by ebramanti** [comment]
While I love vanilla Javascript I am tending to agree with @lexi-lambda on this one. If we are committed with Angular, then we really should be considering Typescript vs. Coffeescript rather than Coffee v. ES6.

I've actually done a bit of research on Typescript for Angular, and what I've found is that it also works great for 1.x applications. It gives us ES6 with a nice way to build Angular apps from the ground up, especially when dealing with a RESTful API.

Most of my learning occurred from Sean Hess' presentation at ng-conf 2014. [Here's the video](https://youtu.be/u6TeBM_SC8w), and he has a repo with the entire presentation [here](https://github.com/seanhess/angularjs-typescript).

What are everyone's thoughts about using Typescript to solve the ES6/Coffee dilemma?

**Comment by jdwolk** [comment]
tl;dr: I'm less in favor of any compile-to-JS langs that aren't giving us serious bang for our buck.

Previously, Coffeescript was a huge win because it hid the disgusting bits of JS + made it harder to shoot ourselves in the foot. The price we pay when using CS is an additional layer of indirection -- I hear everyone screaming "SOURCE MAPS", but in practice I don't think we've seen a solution that has worked consistently with the Rails asset pipeline. Someone please correct me if they've worked on an ongoing basis with something I'm not aware of.

Now that ES6 is more on par w/ what CS offers I, like many of you, see less of a reason to stay with it, especially given the additional price of debugging + that the status quo outside of Railsland is JS and not CS.

Now why not TypeScript?

While I applaud the effort to introduce types into the typeless-cesspool-that-is-JS, I don't think it's worth the cost either. The motivating factor I hear is that it works nicely with Angular 2, but should we pigeonhole ourselves into a compile-to-JS-lang just so we can use a framework which might be yanked by Google on a whim?

I would seriously reconsider TypeScript or any other compile-to-JS-lang only if it satisfies the following criteria:
- Extremely easy to integrate w/ Rails (including source map generation), at least until we foresee ourselves removing JS from Rails projects entirely
- Minimal friction when working with existing JS libs/frameworks
- Has demonstrable longterm support from an active open source community (i.e. not just Microsoft and the Angular core team)
- Offers serious wins, i.e. higher-level abstractions, not just syntactic sugar to hide messier bits of JS

I'm clearly no fan of JS, I'm just not sure TypeScript is _the_ solution. Feel free to disagree / prove me wrong.

`</jds-two-cents>`

**Comment by lexi-lambda** [comment]
@jdwolk: I mostly agree wrt TypeScript—I don't think it makes sense to use _yet_. Maybe we can re-evaluate if we ever start using Angular 2.

I think your points are well-put, and I don't really disagree, but I'd like to add some additional information related to what you brought up.

>  I hear everyone screaming "SOURCE MAPS", but in practice I don't think we've seen a solution that has worked consistently with the Rails asset pipeline.

Yep, this is currently unsupported. However, Sprockets 4, which will ship with Rails 5, [supports source maps](https://github.com/sstephenson/sprockets/pull/632), so theoretically those will be officially supported then.

Additionally, if we start building frontend apps without the asset pipeline, probably using Gulp (we've been using that so far), it obviously has support for source maps out of the box.

> The motivating factor I hear is that it works nicely with Angular 2, but should we pigeonhole ourselves into a compile-to-JS-lang just so we can use a framework which might be yanked by Google on a whim?

I agree with this, but it is probably worth mentioning that Google is allocating a lot of resources to Angular 2, and they appear to be using Angular internally quite extensively, so it's likely that _if_ it catches on (which is, of course, not guaranteed), it will have a long supported lifespan.

> Offers serious wins, i.e. higher-level abstractions, not just syntactic sugar to hide messier bits of JS

CoffeeScript doesn't have this, but TypeScript actually sort of does, which is what Angular 2 is heavily dependent on. Google's AtScript ideas were merged into TypeScript, so now it has `@Annotations`, which do seem to fundamentally change how the Angular 2 team is working with its codebase and designing its APIs. This is why there currently isn't any reasonable JavaScript API for Angular 2—the TypeScript version heavily uses concepts that do not translate to JavaScript in any direct way.

Again, I'm not really advocating for TypeScript (it has other potential problems and shortcomings), but I figured I'd share what I know.

**Comment by patbenatar** [comment]
Hey all - great discussion going on here!

My thoughts:
- We adopt vanilla JS going forward (continue using CS if it's the predominant language in an existing project). Whether a project is using ES5 or ES6 can be determined by the project team, but I suspect a shift to ES6 over time as we get solid transpiler support added to our toolchain + Node 4 supports a good chunk of ES6 out of the box.
- We revisit the TypeScript discussion when Angular 2 drops

**Comment by ebramanti** [comment]
:+1:

**Comment by taboularasa** [comment]
@patbenatar :+1:

**Comment by taboularasa** [comment]
Also :+1: on the kudos to great convo! Good job thinking and expressing our thoughts team! go thinking! yay

**Comment by lexi-lambda** [comment]
Sounds good to me. :+1:

One minor point: ES6 doesn't have list comprehensions, which is a :coffee: feature I use a lot. Babel actually _does_ support them, though, [as an ES7 feature](http://babeljs.io/docs/advanced/transformers/other/comprehensions/). It might be worth looking over the [list of optional extensions](https://babeljs.io/docs/usage/experimental/) and deciding on which ones we should use.

**Comment by jdwolk** [comment]
@lexi-lambda Looks like they're a stage 0 feature, which is still pre-proposal for ES7. I think the "stage 2 or above enabled by default" sounds like a sensible default for Babel and for us to adopt potentially.

Agreed that list comprehensions are a nice feature when langs have them but if they don't, well...use the list monad :trollface: _ducks, awaiting incoming projectiles_

**Comment by lexi-lambda** [comment]
Hmm, yes, you're right. Looks like the current syntax for list comprehensions sucks anyway. :trollface: I'll just use `map` and friends.

**Comment by patbenatar** [comment]
@philosophie/devs Anyone in disagreement with adopting vanilla JS? Speak now or forever hold your peace!

**Comment by keithmerrilliv** [comment]
By the beard of Zeus, let's do it!

**Comment by ebramanti** [comment]
@patbenatar Vanilla JS :heart: :heart: :heart:

**Comment by jdwolk** [comment]
:icecream:

**Comment by patbenatar** [comment]
@nrswolf @mikehmorrissey @wInevitable Everyone in NYC on board with vanilla JS?

**Comment by wInevitable** [comment]
@patbenatar :+1: :+1: :smiley:
